soccer/gameplay/tactics/advance_zone_midfielder.py

```diff
@@ -0,0 +1,138 @@
+                        robocup.Point(0.01, 0.01), 1, 2, \
+                        0.75, 0.5, 50, 1, 0.1)
+
+    class State(enum.Enum):
+        ## getting ready to recieve a pass from another robot
+        passSet = 1
+        ## getting ready to recieve a pass from the dribbling robot backwards for tactical advatage
+        passBack = 2
+
+    def __init__(self):
+        super().__init__(continuous=True)
+
+        #Sends one robot to best reciever point, and other to the best receiving point
+        # of the first midfielder
+        self.add_state(AdvanceZoneMidfielder.State.passSet,
+                       behavior.Behavior.State.running)
+        #To be imblemented later when wanting to play a more conservative 
+        self.add_state(AdvanceZoneMidfielder.State.passBack,
+                       behavior.Behavior.State.running)
+        # old constants in Simple zone midfieler
+        self.y_hold_percent = .8
+        self.x_positions = [-constants.Field.Width / 3, \
+                            constants.Field.Width / 3]
+        self.x_positions_alternate = [-constants.Field.Width/2, \
+                                      constants.Field.Width/2]
+        self.names = ['left_mid', 'right_mid']
+        # Which robot would be best to move to first move point
+        self.optimal = 0
+
+        self.moves = [None, None]
+
+        for s in AdvanceZoneMidfielder.State:
+            self.add_state(s, behavior.Behavior.State.running)
+
+        # more conservative play that should be implemented later
+        self.add_transition(
+            AdvanceZoneMidfielder.State.passSet,
+            AdvanceZoneMidfielder.State.passBack,
+            lambda: False,  #not self.should_get_open(),
+            "When dribbler is in danger and not about to shoot")
+
+        #intially should be passSet method
+        self.add_transition(behavior.Behavior.State.start,
+                            AdvanceZoneMidfielder.State.passSet, lambda: True,
+                            "Immediately")
+
+    def execute_passBack(self):
+        target_y = self.y_hold_percent * main.ball().pos.y
+
+        for i in range(2):
+            if (self.moves[i] is None):
+                self.moves[i] = skills.move.Move(
+                    robocup.Point(self.x_positions[i], target_y))
+                self.add_subbehavior(
+                    self.moves[i], self.names[i], required=False, priority=1)
+            else:
+                self.moves[i].pos = robocup.Point(self.x_positions[i],
+                                                  target_y)
+
+    def execute_passSet(self):
+        #gets the best position to travel to for ball reception
+        best_Pos_From_Ball, value = evaluation.passing_positioning.eval_best_receive_point(
+            main.ball().pos,
+            main.our_robots(), AdvanceZoneMidfielder.FIELD_POS_WEIGHTS,
+            AdvanceZoneMidfielder.NELDER_MEAD_ARGS,
+            AdvanceZoneMidfielder.PASSING_WEIGHTS)
```

Comment:
Since this is the same call that we are doing in `adaptive_formation` it might be more efficient to just pass it. There should be an example of this type of passing in the defense.py <-> submissive_defender.py interaction that you can look at

Reply:
It's in the execute and on_enter dribbling state. It's also stored in the class variable self.pass_target. I'll link the line numbers in a second

Reply:
https://github.com/RoboJackets/robocup-software/blob/master/soccer/gameplay/plays/offense/adaptive_formation.py#L249
Here is where it gets calculated when just dribbling.
https://github.com/RoboJackets/robocup-software/blob/master/soccer/gameplay/plays/offense/adaptive_formation.py#L298
Here is where we calculate it when trying to clear the ball